src/rust/runtime/memory/buffer/demibuffer.rs

```diff
@@ -0,0 +1,824 @@
+}
+
+// Check MetaData structure alignment and size at compile time.
+// Note that alignment must be specified via a literal value in #[repr(align(64))] above, so a defined constant can't
+// be used.  So, if the alignment assert is firing, change the value in the align() to match CPU_DATA_CACHE_LINE_SIZE.
+const _: () = assert!(std::mem::align_of::<MetaData>() == arch::CPU_DATA_CACHE_LINE_SIZE);
+const _: () = assert!(std::mem::size_of::<MetaData>() == 2 * arch::CPU_DATA_CACHE_LINE_SIZE);
+
+// MetaData "offload flags".  These exactly mimic those of DPDK MBufs.
+
+// Indicates this MetaData struct doesn't have the actual data directly attached, but rather this MetaData's buf_addr
+// points to another MetaData's directly attached data.
+const METADATA_F_INDIRECT: u64 = 1 << 62;
+
+impl MetaData {
+    // Note on Reference Counts:
+    // Since we are currently single-threaded, there is no need to use atomic operations for refcnt manipulations.
+    // We should rework the implementation of inc_refcnt() and dec_refcnt() to use atomic operations if this changes.
+    // Also, we intentionally don't check for refcnt overflow.  This matches DPDK's behavior, which doesn't check for
+    // reference count overflow either (we're highly unlikely to ever have 2^16 copies of the same data).
+
+    // Increments the reference count and returns the new value.
+    #[inline]
+    pub fn inc_refcnt(&mut self) -> u16 {
+        self.refcnt += 1;
+        self.refcnt
+    }
+
+    // Decrements the reference count and returns the new value.
+    #[inline]
+    pub fn dec_refcnt(&mut self) -> u16 {
+        self.refcnt -= 1;
+        self.refcnt
+    }
+
+    // Gets the MetaData for the last segment in the buffer chain.
+    #[inline]
+    pub fn get_last_segment(&mut self) -> &mut MetaData {
+        let mut md: &mut MetaData = self;
+        while md.next.is_some() {
+            // Safety: The call to as_mut is safe, as the pointer is aligned and dereferenceable, and the MetaData
+            // struct it points to is initialized properly.
+            md = unsafe { md.next.unwrap().as_mut() };
+        }
+        &mut *md
+    }
+}
+
+// DemiBuffer type tags.
+// Since our MetaData structure is 64-byte aligned, the lower 6 bits of a pointer to it are guaranteed to be zero.
+// We currently only use the lower 2 of those bits to hold the type tag.
+#[derive(PartialEq)]
+enum Tag {
+    Heap = 1,
+    #[cfg(feature = "libdpdk")]
+    Dpdk = 2,
+}
+const TAG_MASK: usize = 0x3;
+
+impl BitOr<Tag> for NonZeroUsize {
+    type Output = Self;
+
+    fn bitor(self, rhs: Tag) -> Self::Output {
+        self | rhs as usize
+    }
+}
+
+impl From<usize> for Tag {
+    fn from(tag_value: usize) -> Self {
+        match tag_value {
+            1 => Tag::Heap,
+            #[cfg(feature = "libdpdk")]
+            2 => Tag::Dpdk,
+            _ => panic!("memory corruption in DemiBuffer pointer"),
+        }
+    }
+}
+
+// The DemiBuffer.
+pub struct DemiBuffer {
```

Comment:
We need an `into_raw_parts()` function in order to convert and consume this data structure and provide a pair of pointers for constructing the scatter-gather array.

Reply:
Today, we do need them due to the way that the scatter-gather array code conversion was engineered originally.

I understand that we may not need it in the long run, but then in order to hook-in your major rework on the scatter-gather array layer is required.

Would it be possible that you provide the `into_raw_parts()` now? Then I can work on getting the scatter-gather array layer cleaner, while you use the buffer to get the missing TCP features implemented.

Reply:
I was going to think about this issue when I integrate this to replace Buffer in a near-future PR.

If I understand correctly, you're arguing for an `into_raw_parts()` function that would be specific to the heap allocated DemiBuffer subtype (similar to how `into_mbuf()` is specific to DPDK allocated ones).  I can do this, but would prefer to make that a separate PR.  [Side note: `into_mbuf()` is needed for interfacing with DPDK FFI routines, I believe the sgarray case can be handled without it]

Anyhow, I'd like to make that a separate PR, as I think there may be other things that I need to expose beyond that to interface properly with everything Buffer did (like, does `data_ptr()` need to be public, or is the current `Deref` enough?) and I don't think I'll know that until I actually attempt to replace Buffer.  `DataBuffer`/`DpdkBuffer` didn't expose the same things, and sometimes exposed different ways to do similar things.  I did look at all the places these are used and attempted to provide the necessary functionality, but in case I missed something, I'd prefer to group them in a latter PR.

Reply:
Sure, it is your call, but keep in mind that:

- This is a requirement to plug in your abstraction to the source tree
- The Scatter-Gather Layer will rapidly change in the upcoming weeks to enable LibOS collocation at the PDPIX layer.

Reply:
Yes, I will have this PR done (or at least ready for review) by the time you return from vacation.

As for the Scatter-Gather layer, I'm concerned that the PDPIX layer as specified doesn't provide enough user flexibility and we may need to expose more of the buffer abstraction functionality (like adjust/trim) through to the Scatter-Gather layer than we currently do.  But that's a topic for another discussion.